src/proxy/outbound.rs

```diff
@@ -381,35 +382,68 @@ impl OutboundConnection {
             });
         }
 
-        let us = us.unwrap();
+        let mut mutable_us = us.unwrap();
+        let workload_ip = self
+            .pi
+            .state
+            .load_balance(
+                &mutable_us.workload,
+                &source_workload,
+                self.pi.metrics.clone(),
+            )
+            .await?;
+
         // For case upstream server has enabled waypoint
-        match self.pi.state.fetch_waypoint(us.workload.clone()).await {
+        match self
+            .pi
+            .state
+            .fetch_waypoint(&mutable_us.workload, workload_ip)
+            .await
+        {
             Ok(None) => {} // workload doesn't have a waypoint; this is fine
             Ok(Some(waypoint_us)) => {
                 let waypoint_workload = waypoint_us.workload;
-                let wp_socket_addr =
-                    SocketAddr::new(waypoint_workload.choose_ip()?, waypoint_us.port);
+                let waypoint_ip = self
+                    .pi
+                    .state
+                    .load_balance(
+                        &waypoint_workload,
+                        &source_workload,
+                        self.pi.metrics.clone(),
+                    )
+                    .await?;
+                let wp_socket_addr = SocketAddr::new(waypoint_ip, waypoint_us.port);
                 return Ok(Request {
                     // Always use HBONE here
                     protocol: Protocol::HBONE,
                     source: source_workload,
                     // Use the original VIP, not translated
                     destination: target,
-                    destination_workload: Some(us.workload),
+                    destination_workload: Some(mutable_us.workload),
                     expected_identity: Some(waypoint_workload.identity()),
                     gateway: wp_socket_addr,
                     // Let the client remote know we are on the inbound path.
                     direction: Direction::Inbound,
                     request_type: RequestType::ToServerWaypoint,
-                    upstream_sans: us.sans.clone(),
+                    upstream_sans: mutable_us.sans,
                 });
             }
             // we expected the workload to have a waypoint, but could not find one
             Err(e) => return Err(Error::UnknownWaypoint(e.to_string())),
         }
+
+        let us = match set_gateway_address(&mut mutable_us, workload_ip, self.pi.hbone_port) {
```

Comment:
I think all calls to `fetch_upstream` (either above, or inside of `fetch_waypoint` end up calling `load_balance`. 

If I didn't mix myself up, I think the calls to `load_balance` and `set_gateway` could be inside `fetch_upstream`. Maybe awkward though.